segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/StorageHealthTracker.java

```diff
@@ -0,0 +1,316 @@
+     * @return value set of storage usage in bytes.
+     */
+    long getStorageUsed() {
+        return storageUsed.get();
+    }
+
+    /**
+     * Sets value set of storage usage in bytes.
+     */
+    void setStorageUsed(long used) {
+        storageUsed.set(used);
+    }
+
+    /**
+     * Gets storage percentage used.
+     * @return double between 0 and 100.
+     */
+    double getStorageUsedPercentage() {
+        return (100.0 * storageUsed.get()) / config.getMaxSafeStorageSize();
+    }
+
+    /**
+     * Gets whether storage is unavailable.
+     */
+    boolean isStorageUnavailable() {
+        return isStorageDegraded.get() || isStorageUnavailable.get();
+    }
+
+    /**
+     * Whether this instance is running under the safe mode or not.
+     *
+     * @return True if safe mode, False otherwise.
+     */
+    boolean isSafeMode() {
+        return isStorageFull.get();
+    }
+
+    /**
+     * Throttles a parallel operation if required.
+     * @return A CompletableFuture that will complete after the throttle is applied.
+     */
+    CompletableFuture<Void> throttleParallelOperation() {
+        if (isStorageDegraded.get() || isStorageUnavailable.get() || shouldThrottle.get()) {
+            return delaySupplier.apply(getParallelThrottleDelay());
+        }
+        return CompletableFuture.completedFuture(null);
+    }
+
+    /**
+     * Throttles an exclusive operation if required.
+     * @return A CompletableFuture that will complete after the throttle is applied.
+     */
+    CompletableFuture<Void> throttleExclusiveOperation() {
+        if (isStorageDegraded.get() || isStorageUnavailable.get() || shouldThrottle.get()) {
+            return delaySupplier.apply(getExclusiveThrottleDelay());
+        }
+        return CompletableFuture.completedFuture(null);
+    }
+
+    /**
+     * Throttles garbage collection batch if required.
+     * @return A CompletableFuture that will complete after the throttle is applied.
+     */
+    CompletableFuture<Void> throttleGarbageCollectionBatch() {
+        if (isStorageDegraded.get() || isStorageUnavailable.get() || shouldThrottle.get()) {
+            return delaySupplier.apply(getGarbageCollectionThrottleDelay());
+        }
+        return CompletableFuture.completedFuture(null);
+    }
+
+    Duration getParallelThrottleDelay() {
+        var millis = config.getMinLateThrottleDurationInMillis() + (percentageLate.doubleValue() * (config.getMaxLateThrottleDurationInMillis() - config.getMinLateThrottleDurationInMillis())) / 100;
+        return Duration.ofMillis(Math.round(millis));
+    }
+
+    Duration getExclusiveThrottleDelay() {
+        var millis = config.getMinLateThrottleDurationInMillis() + (percentageLate.doubleValue() * (config.getMaxLateThrottleDurationInMillis() - config.getMinLateThrottleDurationInMillis())) / 100;
+        return Duration.ofMillis(Math.round(millis));
+    }
+
+    Duration getGarbageCollectionThrottleDelay() {
```

Comment:
All these three ( parallel, exclusive, garbage) will have same delays is it?

Reply:
now they are slightly different